File: functions/us_stock_sectors.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-
"""
美股板块表现 - Netlify Function
定时执行：每个交易日 04:30 (UTC+8)
"""

import json
import yfinance as yf
from datetime import datetime
from typing import List, Dict


# 标普 500 的 11 个板块 ETF
SECTOR_ETFS = {
    'XLK': 'Technology',
    'XLV': 'Health Care',
    'XLF': 'Financial',
    'XLY': 'Consumer Cyclical',
    'XLP': 'Consumer Defensive',
    'XLI': 'Industrials',
    'XLU': 'Utilities',
    'XLB': 'Basic Materials',
    'XLRE': 'Real Estate',
    'XLC': 'Communication Svcs',
    'XLE': 'Energy',
}


def get_sector_performance() -> List[Dict]:
    """获取美股板块表现数据"""
    sectors = []
    
    for symbol, name in SECTOR_ETFS.items():
        try:
            ticker = yf.Ticker(symbol)
            hist = ticker.history(period='1d')
            
            if len(hist) > 0:
                close = float(hist['Close'].iloc[-1])
                open_price = float(hist['Open'].iloc[-1])
                change = close - open_price
                change_pct = (change / open_price) * 100
                volume = int(hist['Volume'].iloc[-1])
                
                sectors.append({
                    'symbol': symbol,
                    'name': name,
                    'price': round(close, 2),
                    'change_pct': round(change_pct, 2),
                    'volume': volume,
                })
        except Exception as e:
            print(f"{symbol} 失败：{e}")
    
    return sectors


def handler(event, context):
    """Netlify Function 入口"""
    print(f"执行时间：{datetime.now().strftime('%Y-%m-%d %H:%M:%S')}")
    
    # 获取数据
    sectors = get_sector_performance()
    
    if not sectors:
        return {
            "statusCode": 500,
            "body": json.dumps({"error": "获取数据失败"}, ensure_ascii=False)
        }
    
    # 按涨跌幅排序
    sectors_sorted = sorted(sectors, key=lambda x: x['change_pct'], reverse=True)
    
    # 生成排名数据
    ranking = []
    for i, s in enumerate(sectors_sorted, 1):
        ranking.append({
            'rank': i,
            'symbol': s['symbol'],
            'name': s['name'],
            'price': s['price'],
            'change_pct': s['change_pct']
        })
    
    # 生成输出
    output = {
        'date': datetime.now().strftime('%Y-%m-%d %H:%M'),
        'market': '美股',
        'ranking': ranking
    }
    
    # 返回 JSON
    return {
        "statusCode": 200,
        "body": json.dumps(output, ensure_ascii=False, indent=2),
        "headers": {
            "Content-Type": "application/json; charset=utf-8"
        }
    }


if __name__ == "__main__":
    # 本地测试
    result = handler({}, {})
    print(result["body"])

File: functions/us_stock_sectors.js
const axios = require('axios');

const SECTOR_ETFS = {
  'XLK': 'Technology',
  'XLV': 'Health Care',
  'XLF': 'Financial',
  'XLY': 'Consumer Cyclical',
  'XLP': 'Consumer Defensive',
  'XLI': 'Industrials',
  'XLU': 'Utilities',
  'XLB': 'Basic Materials',
  'XLRE': 'Real Estate',
  'XLC': 'Communication Svcs',
  'XLE': 'Energy'
};

exports.handler = async (event, context) => {
  try {
    const sectors = [];

    // 获取每个板块 ETF 的数据
    for (const [symbol, name] of Object.entries(SECTOR_ETFS)) {
      try {
        const res = await axios.get(`https://query1.finance.yahoo.com/v8/finance/chart/${symbol}?interval=1d&range=1d`, { 
          headers: { 'User-Agent': 'Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36' } 
        });
        const m = res.data.chart.result[0]?.meta;
        if (m) {
          const open = m.previousClose || m.chartPreviousClose;
          const close = m.regularMarketPrice || m.previousClose;
          const changePct = open ? (((close - open) / open) * 100) : 0;
          const volume = m.regularMarketVolume || 0;
          
          // 估算资金流（成交量 * 平均价格）= 美元，不转换
          const estimatedFlow = volume * close;
          
          sectors.push({ 
            symbol: symbol, 
            name: name, 
            net_money: parseFloat(estimatedFlow.toFixed(2)),
            change_pct: parseFloat(changePct.toFixed(2)),
            volume: volume,
            price: parseFloat(close.toFixed(2))
          });
        }
      } catch(e) {
        console.error(`${symbol} failed:`, e.message);
      }
    }

    // 按净流入排序
    const sorted = sectors.sort((a,b) => b.net_money - a.net_money);

    const ranking = sorted.map((s,i) => ({ 
      rank: i+1, 
      symbol: s.symbol,
      name: s.name, 
      net_money: s.net_money,
      change_pct: s.change_pct 
    }));

    return { 
      statusCode: 200, 
      body: JSON.stringify({ 
        date: new Date().toISOString(), 
        market: '美股', 
        ranking,
        unit: '美元 (ETF 成交额估算)',
        note: '数据基于板块 ETF 成交额估算'
      }, null, 2),
      headers: {
        'Content-Type': 'application/json; charset=utf-8',
        'Access-Control-Allow-Origin': '*'
      }
    };
  } catch(e) {
    return { 
      statusCode: 500, 
      body: JSON.stringify({ error: e.message }),
      headers: { 'Content-Type': 'application/json' }
    };
  }
};
